add apFirst and apSecond

zhangchiqing opened the pull request:
Hi @davidchambers ,

I just moved the changes from https://github.com/sanctuary-js/sanctuary/pull/270#issuecomment-258389389 to here.

I didn't change the function names, as they are still in discussion.

Please review. 

I will create another PR to https://github.com/sanctuary-js/sanctuary/pull/216 once this PR has been merged.

davidchambers commented on README.md:
```diff
@@ -50,7 +50,7 @@ This project provides:
 ## API
 
-<h4 name="TypeClass"><code><a href="https://github.com/sanctuary-js/sanctuary-type-classes/blob/v1.0.0/index.js#L105">TypeClass :: (String, Array TypeClass, a -> Boolean) -> TypeClass</a></code></h4>
+<h4 name="TypeClass"><code><a href="https://github.com/sanctuary-js/sanctuary-type-classes/blob/v1.0.0/index.js#L112">TypeClass :: (String, Array TypeClass, a -> Boolean) -> TypeClass</a></code></h4>
```
As mentioned in the [__CONTRIBUTING.md__][1], it's not necessary to update __README.md__ manually. ;)


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/blob/master/CONTRIBUTING.md

davidchambers commented on index.js:
```diff
@@ -88,6 +88,13 @@
   var identity = function(x) { return x; };
 
+  // kCombinator :: a -> b -> a
+  var kCombinator = function(x) {
```
Let's name this `constant` for consistency with `identity`.

davidchambers commented on index.js:
```diff
@@ -1056,6 +1063,42 @@
+  //# apFirst :: Apply f => f a -> f b -> f a
+  //.
+  //. Combine two effectful actions, keeping only the result of the first.
+  //. Equivalent to Haskell's <* function.
```
Let's write `` `(<*)` `` rather than `<*`.

davidchambers commented on index.js:
````diff
@@ -1056,6 +1063,42 @@
+  //. > apFirst([1, 2], [3, 4])
+  //. [1, 1, 2, 2]
+  //.
+  //. ```javascript
````
This needs to move up a few lines.

zhangchiqing commented:
Thanks for the review @davidchambers .

I've updated the PR for all your comments. And reverted the changes to README.md.

Please let me know they are good :)

davidchambers commented on index.js:
```diff
@@ -88,6 +88,13 @@
   var identity = function(x) { return x; };
 
+  // constant :: a -> b -> a
+  var constant = function(x) {
```
Please add one more space between `//` and `constant` to align the function name in type line with the identifier on the following line. :)

Also, for lack of a better system the helper functions are defined in alphabetical order, so `constant` should be defined before `identity`.

davidchambers commented on index.js:
```diff
@@ -1056,6 +1063,42 @@
+  //# apFirst :: Apply f => (f a, f b) -> f a
+  //.
+  //. Combine two effectful actions, keeping only the result of the first.
+  //. Equivalent to Haskell's (<*) function.
```
Let's wrap the function name in backticks so it will be rendered as "`(<*)`" rather than "(<*)".

zhangchiqing commented:
Good advices! 

Just updated, please take a look again, thanks @davidchambers

davidchambers commented:
Thanks very much for submitting this pull request, @zhangchiqing, and for updating it quickly in response to my suggestions.

I'd like to leave this open for a day or two while we decide on the most appropriate names, but otherwise this looks ready to merge. :)

zhangchiqing commented:
Good to know :)

Yes, I agree. Please let me know if we need to rename them.

Thanks again for the code review.

safareli commented:
@scott-christopher I agree left/right could be confusing because of Either type, in that case I think `first/last` is  better option.

davidchambers commented:
Are you happy with this pull request as it stands, @safareli?

safareli commented:
I would like if we use `apLast` instead of `apSecond`, but if not it's not that big issue.

davidchambers commented:
They're essentially binary operators, aren't they? I think of the blanks in `_ *> _` as _left-hand operand_ and _right-hand operand_, or as _first operand_ and _second operand_. The word _last_ suggests to me we're dealing with a variable number of things rather than exactly two things.

safareli commented:
Agree!

davidchambers commented on index.js:
```diff
@@ -1272,6 +1315,8 @@
     lift2: lift2,
     lift3: lift3,
+    apFirst: apFirst,
+    apSecond: apSecond,
```
One last thing: I think `apFirst` and `apSecond` should come between `ap` and `lift`. Could you move the definitions, the exports, and the tests?

safareli replied:
I think you can still have access to lift2 if you move apFirst because of js hoisting

davidchambers replied:
Good point, @zhangchiqing. The location you chose is perfect! I'll merge this when I get home. :)

davidchambers commented:
Thanks again for this nice contribution, @zhangchiqing. Don't worry about creating a Sanctuary pull request; I will update sanctuary-js/sanctuary#216 to include `S.apFirst` and `S.apSecond`.

:deciduous_tree:

zhangchiqing commented:
Awesome! I'm more than happy to contribute :)

zhangchiqing commented:
One more question:

Will the [README.md](https://github.com/sanctuary-js/sanctuary-type-classes/blob/master/README.md) be updated with the new functions once this PR gets merged? 

Should we run `make all` manually?

safareli commented:
readme is updated when version is bumped

davidchambers commented:
@zhangchiqing, see https://github.com/sanctuary-js/sanctuary-set/pull/1#issuecomment-258782119 if you'd like to know how and when the readme is updated. :)
